providers/flagd/pkg/service/client.go

```diff
@@ -1,73 +1,104 @@
 package service
 
 import (
-	schemaConnectV1 "buf.build/gen/go/open-feature/flagd/bufbuild/connect-go/schema/v1/schemav1connect"
 	"context"
 	"crypto/tls"
 	"crypto/x509"
 	"fmt"
+	"github.com/bufbuild/connect-go"
 	"log"
 	"net"
 	"net/http"
 	"os"
+	"sync"
+
+	schemaConnectV1 "buf.build/gen/go/open-feature/flagd/bufbuild/connect-go/schema/v1/schemav1connect"
+	otelconnect "github.com/bufbuild/connect-opentelemetry-go"
 )
 
 type iClient interface {
 	Instance() schemaConnectV1.ServiceClient
-	Configuration() *ServiceConfiguration
+	Configuration() *Configuration
+}
+
+type Configuration struct {
+	Port            uint16
+	Host            string
+	CertificatePath string
+	SocketPath      string
+	TLSEnabled      bool
+	OtelInterceptor bool
 }
 
 type Client struct {
 	client               schemaConnectV1.ServiceClient
-	ServiceConfiguration *ServiceConfiguration
+	ServiceConfiguration *Configuration
+
+	mu sync.Mutex
 }
 
 // Instance returns an instance of schemaConnectV1.ServiceClient
 func (c *Client) Instance() schemaConnectV1.ServiceClient {
-	if c.client == nil {
-		var dialContext func(ctx context.Context, network string, addr string) (net.Conn, error)
-		var tlsConfig *tls.Config
-		url := fmt.Sprintf("http://%s:%d", c.ServiceConfiguration.Host, c.ServiceConfiguration.Port)
-		// socket
-		if c.ServiceConfiguration.SocketPath != "" {
-			dialContext = func(_ context.Context, _, _ string) (net.Conn, error) {
-				return net.Dial("unix", c.ServiceConfiguration.SocketPath)
-			}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.client != nil {
+		return c.client
+	}
+
```

Comment:
I _think_ we can do the nil check first, and only lock if `c.client` is nil while we create it. That way, we only bother doing any locking if the `c.client` has not been created. Setting it is atomic, I think... so once it's been set and we unlock I guess we can never worry about locking again.

As is, this will lock with every access, which might not be needed. What do you think?

Reply:
After a second look at the code, I think the lock is an overkill. I will try to refactor the code to be bit nicer

Reply:
Refactored - no more locks and null checks. We have NewClient to generate the client [1]

[1] - https://github.com/open-feature/go-sdk-contrib/blob/9dc1bc714da8235dfb77a4e8934e0e6661d38e8d/providers/flagd/pkg/service/client.go